### PR TITLE
Fix match statement giving compiler crash in side the do clause

### LINKED ISSUE #40200: [Bug]: Crash match-stmt in side do-clause
### Description

$subject

### Steps to Reproduce

```ballerina
public function main() {
    var input = [{name: "Saman", price1: [11, 12]}, 
                    {name: "Saman", price1: [19, 20]}];
    _ = from var {name, price1} in input
                do {
                    match name {
                        var a => {}
                    }
                };
}
```

### Affected Version(s)

2201.4.1

### OS, DB, other environment details and versions

_No response_

### Related area

-> Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
$subject

Fixes #40200

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java [hunk 1]
```diff
@@ -2499,7 +2500,13 @@ public void visit(BLangMatchStatement matchStmt) {
     public void visit(BLangMatchClause matchClause) {
         matchClause.matchPatterns.forEach(this::acceptNode);
         this.acceptNode(matchClause.matchGuard);
+        matchClause.declaredVars.forEach((k, v) -> identifiers.putIfAbsent(k, v));
```
Comment (pcnfernando) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java:2503]: Even though shadowing is disallowed; let's not do this. Instead, save the initial `identifiers` and revert back to them as you exit the visit.
Also, you should **not** add only if it's a new symbol. Do `putAll` instead

File: tests/jballerina-unit-test/src/test/resources/test-src/query/match-stmt-in-do-clause.bal [hunk 2]
```diff
@@ -118,13 +118,82 @@ function testConstMatchPattern2() {
...
+                var error(ERROR, code = code) => {
+                    result.push(error(ERROR, code = code));
+                }
+                
+                _ if item is anydata => {
+                    result.push(item);
+                }
+            }
+        };
+    foreach int i in 0...result.length() - 1 {
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/resources/test-src/query/match-stmt-in-do-clause.bal:179]: Ideally you should not loop and check, pass both expected and actual arrays

File: tests/jballerina-unit-test/src/test/resources/test-src/query/match-stmt-in-do-clause.bal [hunk 3]
```diff
@@ -118,13 +118,82 @@ function testConstMatchPattern2() {
...
+
+                var {t: {a, b}} => {
+                    result.push({t: {a, b}});
+                }
+                
+                var error(ERROR, code = code) => {
+                    result.push(error(ERROR, code = code));
+                }
+                
+                _ if item is anydata => {
```
[the remark this answers is not in the document]
Comment (LakshanWeerasinghe) [reply tests/jballerina-unit-test/src/test/resources/test-src/query/match-stmt-in-do-clause.bal:179]: Fixed.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java (hunk 1 above)
Comment (LakshanWeerasinghe) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/QueryDesugar.java]: Updated in [#bf1e4645](https://github.com/ballerina-platform/ballerina-lang/pull/41045/commits/bf1e4645085535ef06fcffc22a51274788e43739).

**Review by pcnfernando — APPROVED** [review]

**Pull request merged**